lib/client/mfa/cli.go

```diff
@@ -0,0 +1,211 @@
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/api/utils/prompt"
+	wancli "github.com/gravitational/teleport/lib/auth/webauthncli"
+	wantypes "github.com/gravitational/teleport/lib/auth/webauthntypes"
+	"github.com/gravitational/teleport/lib/auth/webauthnwin"
+)
+
+// CLIPrompt is the default CLI mfa prompt.
+type CLIPrompt struct {
+	cfg    PromptConfig
+	writer io.Writer
+}
+
+// NewCLIPrompt returns a new CLI mfa prompt with the config and writer.
+func NewCLIPrompt(cfg *PromptConfig, writer io.Writer) *CLIPrompt {
+	return &CLIPrompt{
+		cfg:    *cfg,
+		writer: writer,
+	}
+}
+
+// Run prompts the user to complete an MFA authentication challenge.
+func (c *CLIPrompt) Run(ctx context.Context, chal *proto.MFAAuthenticateChallenge) (*proto.MFAAuthenticateResponse, error) {
+	runOpts, err := c.cfg.GetRunOptions(ctx, chal)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	type response struct {
+		kind string
+		resp *proto.MFAAuthenticateResponse
+		err  error
+	}
+	respC := make(chan response, 2)
+	var wg sync.WaitGroup
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer func() {
+		cancel()
+		// wait for all goroutines to complete to ensure there are no leaks.
+		wg.Wait()
+	}()
+
+	// Use variables below to cancel OTP reads and make sure the goroutine exited.
+	otpCtx, otpCancel := context.WithCancel(ctx)
+	defer otpCancel()
+	otpDone := make(chan struct{})
+	otpCancelAndWait := func() {
+		otpCancel()
+		<-otpDone
+	}
+
+	// Fire TOTP goroutine.
+	if runOpts.promptTOTP {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			defer close(otpDone)
+
+			// Let Webauthn take the prompt below.
+			resp, err := c.promptTOTP(otpCtx, chal, true /*quiet*/)
+			respC <- response{kind: "TOTP", resp: resp, err: err}
+		}()
+	}
+
+	// Fire Webauthn goroutine.
+	if runOpts.promptWebauthn {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			// get webauthn prompt and wrap with otp context handler.
+			prompt := &webauthnPromptWithOTP{
+				LoginPrompt:      c.getWebauthnPrompt(ctx, runOpts.promptTOTP),
+				otpCancelAndWait: otpCancelAndWait,
+			}
+
+			resp, err := c.promptWebauthn(ctx, chal, prompt)
+			respC <- response{kind: "WEBAUTHN", resp: resp, err: err}
+		}()
+	}
+
+	// Wait for the 1-2 authn goroutines above to complete, then close respC.
+	go func() {
+		wg.Wait()
+		close(respC)
+	}()
+
+	// Wait for a successful response, or terminating error, from the 1-2 authn goroutines.
+	// The goroutine above will ensure the response channel is closed once all goroutines are done.
+	for resp := range respC {
+		switch err := resp.err; {
+		case errors.Is(err, wancli.ErrUsingNonRegisteredDevice):
+			// Surface error immediately.
+			return nil, trace.Wrap(resp.err)
+		case err != nil:
+			c.cfg.Log.WithError(err).Debugf("%s authentication failed", resp.kind)
+			// Continue to give the other authn goroutine a chance to succeed.
+			// If both have failed, this will exit the loop.
```

Comment:
If both have failed, won't this exit the loop be returning the default value of `resp.resp, nil`? (If so, is that what we want?)

I'm still a bit lost, so with the caveat that this may just not work -- I wonder if we can make the logic more easily comprehensible by using separate channels for `totpRespC` and `webauthnRespC`, and `select`-ing on them.